### ISSUE TITLE
Add zsh-like reverse-isearch

### ISSUE THREAD

**Issue by ghost** [fish-shell/fish-shell#602]
While I love most things about fish, I had to switch back to zsh because C-r in zsh (or even bash) is more efficient than fish's reverse-search functionality.

In zsh you can type-to-complete substring-matches in history. With fish, you can cycle to see what in your history matches your substring, but you can't easily modify your substring while searching to narrow down the match.

For example if I type "git re" and go up a few times, I may find that I'm only finding a bunch of "git reset"-ish commands, but I know I want a command like "git rebase", so in zsh or bash I would just type a "b" to narrow that down. But in fish I have to restart my search with C-c and try again with "git reb". Otherwise I'm just adding a "b" to the end of whatever the last match was, which isn't what I meant to do.

This may appear trivial, but I rely heavily on C-r for all my commands (which is one reason I like fish, because it predicted that correctly about me), but I rely on it for substring-matches, not just prefix-string matches like fish does with auto-suggestions.

I'm proposing that in addition to the existing reverse-search feature, a zsh-like C-r be implemented as well.

**Comment by kfirshay** [comment]
+++1

**Comment by munum** [comment]
+1

**Comment by numkem** [comment]
+1

Maybe implement what @jbonjean did, works pretty well.

**Comment by simar7** [comment]
Add me to the "me-too" list as well. Thank you for the hardwork!

**Comment by zhen9ao** [comment]
+1

**Comment by pathikrit** [comment]
+1

**Comment by jinchizhong** [comment]
@jbonjean really cool solution!!! thank you very much.

**Comment by dsbonev** [comment]
:1

**Comment by chrismacdermaid** [comment]
+1

**Comment by davidmh** [comment]
For anyone interesed, [FZF](https://github.com/junegunn/fzf) as an amazing implementation of this feature for fish:

https://github.com/junegunn/fzf#key-bindings-for-command-line

![fzf's ctr-r](http://puu.sh/jqGMQ/c034f63c2c.png)

Which, IMO is a great alternative for the people worried about the "weird modal state" and as the huge advantage of the fuzzy search.

**Comment by coldis** [comment]
+1 for FZF

**Comment by derekstavis** [comment]
What about using a completion state for history search? We already have a great UI for it, a very nice search, and up/down arrows.

**Comment by jfindlay** [comment]
To me the main virtue of bash `C-r` style search vs fish history search is that whereas fish history search is anchored to the beginning of the line, with `C-r` you can search for a string starting anywhere in the previous command line.

If you're going back to select a command from among several that all share the same first few tokens (like running a command with a variation of options), using fish's up arrow style search is not going to be as efficient as a `C-r`, where you can navigate immediately to the difference.

**Comment by jakwings** [comment]
Bash's C-r, C-s and C-o are very helpful, much better than fish's simple up/down (or C-p/C-n) arrow search.
C-r does reverse-search, C-s forward, and edit the search anytime
C-p/C-n or Up/Down to the previous or next command in history (stop searching)
C-o to execute the current command, then get the next command in history

These are my most used bash shortcuts. And thank @jbonjean, good work!

**Comment by mfilotto** [comment]
Hi, any update on this one ?

**Comment by faho** [comment]
@mfilotto: No.

**Comment by pickfire** [comment]
@mfilotto, no one is assigned to it.

**Comment by nakaomote** [comment]
My 2 cents is that, the current ctrl+p/n setup is just great. As a long time zsh user I didn't feel at all held back by ctrl+p/n. That said:

ctrl+o (then ctrl+j) is definitely marvelous to have and I use it in both bash and zsh. I would love to see ctrl+o implemented.

**Comment by obfuscated** [comment]
Up-down arrow keys are fine, just please add live editing and all will be fine with this feature.

**Comment by NeMO84** [comment]
@davidmh's suggestion is an excellent alternative to not having the native feel of the crtl+r. And so far I actually like it better. Getting to see more results makes the searching/lookup even more intuitive as I don't have to use my memory for those cases where the lookup text is very dynamic.

**Comment by nordlow** [comment]
Yes, I want this.

**Comment by bmc** [comment]
At the risk of adding a "me, too", the lack of Ctrl-R is my only major complaint about Fish. I can live with the lack of Posix shell compliance, the lack of "!!" and "!$", and the need to remember to type `; and` instead of `&&`. But Ctrl-R is a deal-breaker for me. While the arguments against it all have a certain measure of validity (lack of discoverability, the confusing modal nature of Ctrl-R, etc.), I find that they aren't compelling arguments for me, primarily for one reason: _Many_ other tools that I use (such as the Scala REPL, IPython, `irb`/`pry`, the R REPL, and Emacs, to name a few) support readline-like key bindings, including Ctrl-R. Ctrl-R is wired into my muscle memory at this point. Even though the arguments against Ctrl-R might be compelling, those _other_ tools I use aren't going to change to the Fish way of doing things, and my fingers and brain prefer consistency. For my money, Ctrl-R support in Fish would be a huge win. YMMV.

**Comment by faho** [comment]
[Several "+1" comments have been removed. Keep in mind this issue is open and these somewhat pollute discussion of the issue. If you don't have a substantive or particularly helpful response, adding a :thumbsup: reaction to the issue is a good way to vote. —The Management]

**Comment by andrewrk** [comment]
> What is it you like about ctrl-r, that isn't served by the up-arrow history search?

One thing: You get search results as you type each character. With Fish, you have to type a complete query before pushing up. So you might waste characters when a shorter query would have worked, or you might not type a large enough query, and trying to add more characters to the query accepts the proposed completion and appends the character.

**Comment by Hi-Angel** [comment]
@ridiculousfish, **@jzanchey**, answering the advantage question: with Fish's prefix search `/mnt/kubuntu/\@/opt/google/chrome/google-chrome --enable-native-gpu-memory-buffers` can only be found by the name of root dir, but neither it's a single command from the dir, nor it is comfortable to recall, was it `/media`, `/mnt`, or perhaps even `/opt`, or anything else. *Reverse-search* allows to type any characters from the memory to find the command.

A bad example? Okay, what about `vim` history: I recently installed Archlinux, and editing around configs a lot. So, I can't just type in Fish `vim /`, and scroll up to a filename, it'd be faster to type a path manually. However, instead I *reverse-search* part of the filename *(having a lot less false positives)*, quickly getting the command at hands.

It have nothing to do with "muscle memory". Fish is amazing, and if, for whatever reason, it has another keybinding, I could sacrifice the habit. But its absence forced me to search for zsh Fish-like configs; and only its fail forced me to read the comments, and to discover [jbonjean made a workaround](https://github.com/fish-shell/fish-shell/issues/602#issuecomment-45051065). 

I do use r-search, probably, every third command. I think, you do understand now, why people says it's a showstopper.

**Comment by terrycojones** [comment]
With C-r in bash you can also hit backspace to remove a character from the current search prefix, and the interactive search continues.

I miss C-r so much :-(

While we're on the subject of things we miss, does fish have an alternate for `[abc]` (i.e., match any single `a`, `b`, or `c` char), or for `?`  I also miss being able to do a C-l in the middle of a search in order to clear the screen. I can do it in fish, but that stops the search.

Sorry if this sounds like whining. Fish is fantastic, and it shows a TON of innovative thinking - it's really impressive. Maybe everyone has a few things they really miss from other shells. C-r certainly seems to be one.

**Comment by bassu** [comment]
Well, there's a reason why this issue is still open 😜 
 _So long_ ...  🐟 !

**Comment by andrewrk** [comment]
I think fish should pretty much steal everything that fzf does and make it part of fish. In the meantime, having both installed works fine.

**Comment by peduarte** [comment]
Nothing yet?! hahahaha brilliant!

**Comment by dysinger** [comment]
C-r was the reason I had to go back to zsh also

**Comment by iampuma** [comment]
After trying out Fish, this is actually the only functionality that really makes me not switch to Fish. So hopefully this will be really considered as an included feature out of the box. 

I would even go so far to say that the i-search feature should be extending Fish completion to be the second step in completing the user command if Fish does not show a HIT. So even pressing C-r would be redundant.

**Comment by dysinger** [comment]
I commented a week or two ago but it seems to be gone. I said that I tried fish several times only to go back to zsh because of reverse search not being full text like zsh. I'm reaffirming what many many others have said on their own. It's not just a :+1:.

**Comment by ridiculousfish** [comment]
@dysinger What do you mean by "full text?"

**Comment by dysinger** [comment]
I press CTRL-R and I type any word and zsh pulls up matches from my history.  Fish requires that I type the beginning of the line.  An example in screenshots:

![selection_252](https://cloud.githubusercontent.com/assets/447/23534133/9abc252c-ff6a-11e6-9460-6a2ca7ceb989.png)
![selection_251](https://cloud.githubusercontent.com/assets/447/23534134/9acc235a-ff6a-11e6-91bc-3b2b0fa81023.png)

**Comment by krader1961** [comment]
> Fish requires that I type the beginning of the line.

Not true unless I've misunderstood you. Type some text that appears anywhere in a previous command then press [up-arrow] or [ctrl-P]. It even honors whitespace.

**Comment by jordansamuels** [comment]
@krader1961 while you are correct that you can search for text that appears anywhere in the line, the ergonomics are inferior to those of true reverse i-search functionality.  For example, let's say your history contains  `echo foo taz` and you want to change it to `echo bar taz`.  In zsh, you can type `CTRL-R foo` then you will have the cursor **placed before the word `foo`**.  This placement is critical.  Then you can just delete the `foo` and replace with `bar`.  Unless I'm missing something, in fish, you type `foo` then [ctrl-P], but then your cursor is at the **end of the line**, which means you still need to navigate all the way back to `foo` in order to replace it.  This imposes a disproportionate tax on keystrokes required.  With the zsh way, the number of keystrokes required is constant as a function of line length; with fish it is not.

This may seem like nitpicking, but for me it's actually critical.  I've been trying out fish for about 6 months and I love everything else about it.  I'm reluctant to switch back to zsh, but unless there's a more efficient way to edit command lines, I'll probably have to, since for me efficiency trumps everything else.

**Comment by ridiculousfish** [comment]
@jordansamuels thank you for your thoughtful writeup. Be assured it's not at all nit-picky. Thank you also to everyone who thumbs-uped this, that's valuable feedback too.

If fish positioned the cursor at the relevant token, would that address your concerns? That seems like a completely reasonable change to me.

Also zsh-newbie question - in your example, what keystrokes do you use to 1. exit from history search mode, and 2. delete the `foo` text? I would expect it would be simpler to position the cursor at the end of the match instead of the beginning, because then a ctrl-w can delete the token; what am I missing?

**Comment by jordansamuels** [comment]
@ridiculousfish I would say yes, if fish positioned the cursor at the relevant token, it would address enough of my concerns that I'd consider fish a big win at this point.  I can't speak for the other people on this thread though.  And of course, I'll probably notice other differences at some point; but my best guess is that this is "good enough".

As for your questions: 1. to exit from history search mode, I use ctrl-G or esc-esc (which seems to be needed in some setups).  2. To delete the `foo` text, I use meta-d (or esc, d) which is delete-word.  These are standard emacs key bindings and also used by the zsh line editor [zle](http://zsh.sourceforge.net/Doc/Release/Zsh-Line-Editor.html).  Also note that any navigation command such as ctrl-f/ctrl-b for forward/backward will also exit from the search mode.  

As for positioning at end of match instead of beginning, in theory you may be correct; but in practice, the muscle memory of all the zsh users comes from emacs, which has positioned the cursor at the *beginning* of the isearch term since time immemorial.  So not just for myself, but I'm guessing for many of the other prospective converts from zsh, placing at beginning is superior (but again, I can't speak directly for them, only for me).

Thanks,
Jordan

**Comment by terrycojones** [comment]
Just to add a +1 to Jordan's comments.  Any navigation or meta command
stops the search. Cursor at start of matched token. I also use M-d to
delete the word. I'm not coming from zsh, but from bash and have also been
using emacs for some decades :-)    Thanks for still considering this
change. I miss C-r every day but I still stick with fish, which has a crazy
amount of innovation in something as old as the shell.  And, seeing as
there are so many of us missing this so much, you'd think we might have
"fixed" it ourselves by now...

**Comment by jordansamuels** [comment]
@ridiculousfish any movement on this?  I might be inclined to help code it if necessary.

**Comment by faho** [comment]
@jordansamuels: I'm not aware of any. And if someone is working on anything, they should comment in the corresponding issue.

So feel free to give it a try!

**Comment by braham-snyder** [comment]
@dysinger @Hi-Angel Just to clarify a potentially confusing point: the faint, gray text you see following your current input (`-f /usr/bin/fish ...` in your screenshot @dysinger) is a prefix-only ["autosuggestion"](https://fishshell.com/docs/current/tutorial.html#tut_autosuggestions), completely distinct from the full-line search available with the up/down arrow keys. The up/down arrow keys will indeed match, e.g.,`test` to a previous command `echo "this is a test"`, even when the autosuggestion shown is different (autosuggestions are accepted with different shortcuts--see that prior link).

That said, I too highly recommend the fish-compatible [fzf](https://github.com/junegunn/fzf#key-bindings-for-command-line) for its incremental command history search.

**Comment by sametmax** [comment]
No news may, so I'm upping this. 

It's not just that ctrl + r is superior to the history.

It's also that when you move between a lot of machines, some have bash, some have zsh, and some have fish. Obviously the muscle memory kicks in, and you end up very frustrated.

**Comment by Hi-Angel** [comment]
/me migrated to zsh long ago, no regrets. It does require setting up, but otherwise is of almost the same level as fish *(some trivia was missing, but can remember what, something unimportant though)*. Lack of reverse-search is not the only problem, fish also has some bash-incompatibilty causing occasional code from StackOverflow to fail. zsh is way more tolerating in that regard.

**Comment by giodamelio** [comment]
In the meantime I've worked around this pretty well with.

```
function reverse_history_search
  history | fzf --no-sort | read -l command
  if test $command
    commandline -rb $command
  end
end
bind \cr reverse_history_search 
```

P.S. Anyone know why `test -n $command` did not work as I expected it to?

**Comment by AlbertVeli** [comment]
I installed fzf today to try it out. It automatically added a keybinding for C-r without any need for writing a custom fish function, effectively working around the lacking C-r support in fish.

**Comment by Hi-Angel** [comment]
> Interestingly enough, I want the fish functionality in zsh. ￼😂

@NightMachinary which one? I haven't used fish for a long time, but AFAIR oh-my-zsh + [fuzzy-completion](https://superuser.com/a/815317/304864#does-a-fuzzy-matching-mode-exist-for-the-zsh-shell) gets very close to it *(some trivia was different)*.

**Comment by NightMachinary** [comment]
Cycle between completions with typed prefix. Zsh actually had it, but it is
buggy and didn’t work on my system. I found a workaround though and now I’m
happy ^_^

**Comment by arigit** [comment]
Discovered fish just last week after 15 years of bash and a) immediately loved it for the most part b) immediately hated that there was no ctrl-r partial searching. This made me land into this bug and discover fzf (debian/ubuntu users: there is no official package yet, but it's trivial to build following the instructions in github: https://github.com/junegunn/fzf )

As said above, hoping that all the functionality of fzf will be incorporated into fish to make the fish shell experience seamless. Without this functionality, fish is not a true complete modern shell solution, it's missing part of one leg

**Comment by suprafly** [comment]
I concur, I just _attempted_ the switch but the lack of `Ctrl + R` support is killing me. From a user experience perspective, the shell promises sensible defaults but does not support a default that seems sensible when arriving from bash. That's nonsensical.

I want to love fish.
Help computer.

**Comment by suprafly** [comment]
Work around,

[Install fzf to give you the recursive search](https://github.com/junegunn/fzf)

[Install the fish package manager, fisherman](https://github.com/jorgebucaran/fisher)

[Then install the fish fzf plugin for fisherman to get the keybindings working in fish](https://github.com/jethrokuan/fzf)

But I still like @eskhool 's recommendation most of all.

**Comment by myrdd** [comment]
Just wanted to add to the discussion that I also really love this zsh key binding:

```
history-incremental-pattern-search-backward
```

It allows for backwards glob search:

<pre>
% <b>vi ~/.zsh</b>rc
bck-i-search: vi*zsh_
</pre>

and using `^` for ”command starts with“ (regex-like):

<pre>
% <b>ba</b>sh                                 
bck-i-search: ^ba_         
</pre>

**Comment by fdr** [comment]
I don't use zsh, but I do use other readline programs, like `pry` or `psql`. A bit of consistency would be nice.

**Comment by maestrofx** [comment]
A little bit annoying, but, it's ok then. As the autosuggestion is a cool feature, type part of `history_command` you remember, do up/down arrow to see that's the command you want. It's enough!

**Comment by maraoz** [comment]
For anyone arriving here in ~Feb/2019, the best solution is still installing [fzf](https://github.com/junegunn/fzf) as far as i can tell.
Hope I saved you 30' of research/reading.

![screenshot from 2019-02-14 11-33-05](https://user-images.githubusercontent.com/287189/52793387-5f896b80-304c-11e9-9570-1eaa5d17c7a5.png) 
<3

**Comment by emorozov** [comment]
@maraoz Unfortunately, `fzf` is not a replacement for the incremental search. It search in a weird way, I spent much more time finding the right command in `fzf` compared to the search in bash/zsh. Fzf tends to put incorrect results first for some reason. Also, it feels weird, and you don't see what was on the screen before you called fzf.

All things considered, it's not a solution, and actually for me `fzf` is even more annoying than no search at all. I'll have to switch back to bash/zsh until fish finally implements this feature...

**Comment by danielb2** [comment]
@emorozov it doesn't sound like you're using this: https://github.com/junegunn/fzf/blob/master/shell/key-bindings.zsh

fzf just uses fuzzy searching, and with the keybindings it's a cinch. just ctrl+r like you're used to, and start typing in your inc search. You can be very sloppy since it's fuzzy.

**Comment by emorozov** [comment]
I use this solution I've found on the net:
```
function reverse_history_search
  history | fzf --no-sort | read -l command
  if test $command
    commandline -rb $command
  end
end

function fish_user_key_bindings
  bind \cr reverse_history_search
end
```

Maybe I have to tweak it, but in it's current form is highly annoying because it's nowhere as convenient as traditional bash/zsh search. I've tried to get used to fish autocompletion for more than 6 months, but it is still a huge pain point for me. I'm far less efficient with it than I was in bash/zsh.

**Comment by danielb2** [comment]
> All things considered, it's not a solution, and actually for me `fzf` is even more annoying than no search at all. I'll have to switch back to bash/zsh until fish finally implements this feature...

So you can try https://github.com/jbonjean/re-search instead.

**Comment by danielb2** [comment]
@emorozov  the other thing you can do with fzf is prefix your search with `'` to get to an exact-match similar to what you're used to.

**Comment by danielb2** [comment]
@emorozov the other thing to tweak is to use this: https://github.com/junegunn/fzf/blob/master/shell/key-bindings.fish

and if you like the exact search by default, you can modify the file:

```
     if [ $FZF_TMUX -eq 1 ]
       if set -q FZF_TMUX_HEIGHT
-        echo "fzf-tmux -d$FZF_TMUX_HEIGHT"
+        echo "fzf-tmux --exact -d$FZF_TMUX_HEIGHT"
       else
-        echo "fzf-tmux -d40%"
+        echo "fzf-tmux --exact -d40%"
       end
     else
-      echo "fzf"
+      echo "fzf --exact"
     end
   end
```

**Comment by dgutov** [comment]
Regarding the fzf integrations, does anybody have a solution to make `Ctrl-C`abort the command instead of having it linger in `read -l command`?

**Comment by danielb2** [comment]
@dgutov ctrl+c works fine for me.

**Comment by dgutov** [comment]
Please disregard. My problem was that fzf (0.17.1) hides the most recent prompt after Ctrl-C when used with the flag `--height`.

**Comment by monokal** [comment]
Open issue for 6 years, nice one team.

**Comment by nhooyr** [comment]
> Open issue for 6 years, nice one team.

Not cool. This is a project ran by volunteers, no one owes you anything.

**Comment by DestyNova** [comment]
I was just about to invite the commenter to submit a PR...

**Comment by EmpiricEmpire** [comment]
> he deleted his comment.

Oh, don't you worry, internet does not forget. The comment was done by @monokal, 10 November, 2019, at 16:37 by Greenwich. Screenshot:

![Screenshot_20191110_233341](https://user-images.githubusercontent.com/47756298/68550276-8aa1bf80-0412-11ea-846a-0ebf85cc36d9.png)

Don't thank me.

**Comment by faho** [comment]
There was a bit of a guffaw here about someone leaving a rude comment and subsequently deleting it.

I removed everything pertaining to it, and I invite all involved to leave it at that.

**Comment by miguelmota** [comment]
Fish is great and it's almost near damn perfect but lack of `ctrl-r` reverse search functionality made me go back to bash and zsh since it's a feature I use very frequently. 

I see this issue was opened in 2013 and it's already 2020; doesn't seem like fish authors will add this feature anytime soon 🙁

**Comment by meeDamian** [comment]
That was my big issue with `fish` too, until I've discovered that it kinda has the same, although discovery is terrible…

Other shells:
* `ctrl+r`
* type…
* `ctrl+r` to iterate

Fish:
* type…
* `arrow up` until found

**Comment by miguelmota** [comment]
@meeDamian very interesting; will take some getting used to, although using arrow keys is a big dislike to me.

**Comment by terrycojones** [comment]
Using up arrow doesn't result in the same behavior you have in bash. Once you hit up arrow, the entry of search text is finished. I.e., you have to know in advance exactly what to search for, and exactly how much of it to type, before you hit up arrow. Once you hit up arrow you can no longer type additional characters to make the search more specific or hit backspace to remove characters from the search string. It's very awkward, and not at like the flexibility you have with control-r search to keep on adjusting the search text until you find what you want. I've used fish for years now, and miss control-r every day. I should make better use of fzf.

Another thing (different topic) I find awkward with fish is if you hit TAB while you're in the middle of text, it doesn't complete on the text before where your cursor is. It seems to consider the full text string for completion and moves your cursor to the end of the full text. This would presumably be relatively simple to fix.  E.g., if my cursor is on the w in helloworld and I have a file called hello.py in the directory, I would like the TAB to expand the hello before my cursor to hello.py but instead I end up with the helloworld text unchanged and the cursor at the end of the line. That's doubly unhelpful, as it doesn't do the completion AND it moves my cursor away from where I'm trying to edit.

Anyway, no complaints from me! This is all free software, no one has paid a cent, no one is entitled to anything. And it's open source, as had been pointed out. Instead we owe thanks - fish is remarkably innovative in many ways.

**Comment by krobelus** [comment]
@terrycojones 
> I should make better use of fzf.

Both `fzf` and `skim` ship with a fish binding including control-r to search history - call `fzf_key_bindings`.

> E.g., if my cursor is on the w in helloworld and I have a file called hello.py in the directory, I would like the TAB to expand the hello before my cursor to hello.py but instead I end up with the helloworld text unchanged and the cursor at the end of the line. That's doubly unhelpful, as it doesn't do the completion AND it moves my cursor away from where I'm trying to edit.

The cursor movement on failed completion will be fixed in 3.1.0. Your example doesn't work because fish tries to complete the entire token - that's unlikely to change, just add a space.

@miguelmota 
> using arrow keys is a big dislike to me.

use control-p / control-n / control-f / control-b

**Comment by terrycojones** [comment]
@krobelus - thanks for the reply Johannes. Maybe the TAB completion could fall back to trying with an inserted space if the current behavior fails to find a completion? E.g., in my example the helloworld completion fails, so fish tries with hello (the text before my cursor). Or maybe there could be a setting to allow either behavior. TAB for completion (at least to my mind) means "complete the text I've just typed" (i.e., what's to the left of my cursor), not the text that's "ahead" of what I'm now trying to enter. I think completions based on text to the left is arguably more likely to be where the user wants the completion to occur - otherwise why would they have moved the cursor back to that precise location?

BTW, I recently wrote a little (toy?) shell in Python. See https://github.com/terrycojones/daudin in case you're interested. I still use fish, though :-)

**Comment by krobelus** [comment]
> Maybe the TAB completion could fall back to trying with an inserted space if the current behavior fails to find a completion?

I think we can do something here long-term, let's track this in #6687.

> Once you hit up arrow you can no longer type additional characters to make the search more specific

Yeah, that can be slightly inconvenient. This will be fixed by #6686, which is a big step towards incremental search. If anyone wants to try the new behavior and give feedback on whether it works as expected, now is a good time :)

**Comment by Thovthe** [comment]
@krobelus you are making fish 3.2.0 very tasty, keep up the good work.

**Comment by ahmeed83** [comment]
This one did the job! 

https://github.com/jethrokuan/fzf

**Comment by jdz** [comment]
Will the "new" incremental search feature also support wildcards, and will it be possible to switch search direction?  Like with ZSH's `history-incremental-pattern-search-backward` and `history-incremental-pattern-search-forward` functions?

Funny that I started trying Fish yesterday, and this is the feature I need, and it's been controversial since 2013.  Also sad.

**Comment by Nemoden** [comment]
Just dropping by to say I'm missing this feature from bash/zsh, wanted to try `fish` for the first time and it looks and feels great, but because of the lack of Ctrl+R, gotta go back to `zsh`... it's simply a huge habit... well, maybe a habit I can break and get a new one if there was a replacement for ctrl+r in fish

**Comment by psychoslave** [comment]
Interesting to see how the Fish team is staying focus on its philosophy. That's the second topic where I meet it, after the one [on bang](https://github.com/fish-shell/fish-shell/issues/288).

While I understand the point of discoverability, I don't see how using C-p as a alternative of C-r is any better.

On the same level C-f to complete the current line with the suggestion is not more discoverable, you have to read the documentation. I had to check, and it seems that in the defaults, right arrow do the same completion. But it's so less convenient to my habits that I almost forgot it and had to check to see if it works that way. My point is, for many user the right arrow is more discoverable, and that's fine for them. The existence of C-f doesn't hurt this kind of user profile. On the other hand, those who prefer something less distant from base position on the keyboard will be better served with C-f and can forget about the right arrow binding.

Now, sure there are surely way to work around with custom plugins and so on, but given that the C-r is so common in other CLI (not only shells), I don't find especially friendly to have to configure external tools to obtain this basic expectation. Those who don't know C-r wouldn't be hurt by it's existence as a default, and those who are already accustomed with such a feature in CLI would be better served with its default support.

All that said, I must say I appreciate the team philosophy and hope this comment doesn't feel to harsh. Thanks for this wonderful tool.

**Comment by psychoslave** [comment]
Hey, I'm progressing in my reading of the documentation, and I just discovered that `C-s` allow to filter among a set of suggestion. I tried out with `dnf install <tab><Cs>`, and it's amazingly quick and efficient. That's the kind of behaviour that makes me glade that I made the switch to Fish. I wanted to to try if there is a `C-s` in Bash too, but actually just typing `dnf install <tab>` made the prompt freeze. And in Zsh it completed the line as `dnf install 0ad-0.0.23b-9.fc31.x86_64` then freezed. :astonished: Now, Bash and Zsh might let improve the situation with some tweaks, but Fish came out of the box with a great user experience here. All my praise to the Fish team for that.

Now, that make the absence of a `C-r` all the more strange from my perspective, as what `C-s` in this tab completion menu is not more discoverable without reading the manual (but surely it could, see https://github.com/fish-shell/fish-shell/issues/7026).
